libs/ngxtension/linked-query-param/src/linked-query-param.ts

```diff
@@ -0,0 +1,414 @@
+			const value: string | null = params[key] ?? null;
+			// If a parsing function is provided in the config, use it to parse the value.
+			if (options?.parse) {
+				return options.parse(value);
+			}
+			// If the value is undefined or null and a default value is provided, return the default value.
+			if (
+				(value === undefined || value === null) &&
+				options?.defaultValue !== undefined
+			) {
+				return options.defaultValue;
+			}
+			// Otherwise, return the original value or the parsed value (if it was parsed).
+			return value;
+		};
+
+		// create a signal that is updated whenever the query param changes
+		const queryParamValue = toSignal(
+			route.queryParams.pipe(
+				distinctUntilKeyChanged(key), // skip if no changes on same key
+				map((x) => parseParamValue(x)),
+			),
+			{ initialValue: parseParamValue(route.snapshot.queryParams) },
+		);
+
+		const source = signal<T>(queryParamValue() as T);
+
+		const originalSet = source.set;
+
+		effect(() => {
+			const x = queryParamValue();
+			// update the source signal whenever the query param changes
+			untracked(() => originalSet(x as T));
+		});
+
+		const set = (value: T) => {
+			// we first set the initial value so it synchronous (same as a normal signal)
+			originalSet(value);
+
+			// when the source signal changes, update the query param
+			// store the new value in the current keys so that we can coalesce the navigation
+			let valueToBeSet: any = value;
+			if (options?.serialize) {
+				valueToBeSet = options.serialize(value);
+			} else if (value === undefined || value === null) {
+				valueToBeSet = null;
+			} else {
+				valueToBeSet = typeof value === 'string' ? value : String(value);
+			}
+
+			globalHandler.setParamKeyValue(key, valueToBeSet);
+			globalHandler.setCurrentNavigationExtras(options ?? {});
+
+			// schedule the navigation event (multiple synchronous navigations will be coalesced)
+			// this will also reset the current keys and navigation extras after the navigation
+			globalHandler._schedulerNotifier.notify();
+		};
+
+		const update = (fn: (value: T) => T) => set(fn(source()));
+
+		return Object.assign(source, { set, update });
+	});
+}
+
+/**
+ * Can be used to parse a query param value to a number.
+ * You can also use the `defaultValue` option to set a default value if the query param is not present in the url (null or undefined).
+ *
+ * Example:
+ * ```ts
+ * linkedQueryParam('page', { parse: paramToNumber() });
+ * ```
+ * Will return null if the query param is not present in the url.
+ *
+ * Or with a default value:
+ * ```ts
+ * linkedQueryParam('page', { parse: paramToNumber({defaultValue: 1}) });
+ * ```
+ *
+ * Will return 1 if the query param is not present in the url.
+ */
+export function paramToNumber(config: {
+	defaultValue: number;
+}): (x: string | null) => number;
+export function paramToNumber(config?: {
+	defaultValue?: number | null | undefined;
+}): (x: string | null) => number | null;
```

Comment:
question: would `export function paramToNumber(): number | null` work? It looks like as soon as I provide a `config` object, the first overload takes over and I can't provide `null | undefined` to `defaultValue` anyway so why would the second overload needs to declare the optional `config?` object at all?

Reply:
I wanted to have the possibility to have the type number only on the paramToNumber if the user provides a number and remove null as a type that the signal can have.

Reply:
```ts
  _withNumberParse = linkedQueryParam('page', { parse: paramToNumber({defaultValue:1}) });

// will produce this type

//                               👇 you will never get null from the type
_withNumberParse: WritableSignal<number> & {   
   set: SignalSetFn<number | null>   
   update: SignalUpdateFn<number | null> 
 }
```

Reply:
<img width="731" alt="image" src="https://github.com/user-attachments/assets/a7171a81-356c-4792-8451-4bc04afd0965">
you would still keep your overloads but simplify one of the 2 overloads. This is what I meant